### PR TITLE
MM-15020 Adding server side plugin localization framework.

### PR DESCRIPTION
**Pull request by crspeller**
#### Summary
Adds basic functionally for translating server side plugin messages.

Implementation for demo plugin: https://github.com/mattermost/mattermost-plugin-demo/pull/33

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-15020

### CODE REVIEW

**Comment by crspeller** [comment]
@cpoile I don't see any comments on that pr?

**Comment by cpoile** [comment]
@crspeller hah, forgot to hit the final comment button. :) Up now.

File: plugin/helpers_i18n.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/mattermost/go-i18n/i18n"
+	"github.com/mattermost/mattermost-server/model"
+)
+
+func (p *HelpersImpl) TServer(translationID string, args ...interface{}) string {
+	if p.localizationSettings == nil {
+		return translationID
+	}
+
+	if p.locales[*p.localizationSettings.DefaultClientLocale] != "" {
```
Comment (hmhealey) [inline plugin/helpers_i18n.go]: Shouldn't this use the DefaultServerLocale?

**Comment by hmhealey** [comment]
Oh, `ServeHTTP` does include a plugin context. I didn't see that before. Never mind my comment about translating API responses

File: plugin/helpers_i18n.go (hunk 1 above)
Comment (crspeller) [reply plugin/helpers_i18n.go]: From the system console: `Default language for newly created users and pages where the user hasn't logged in.` I think that's what we want. The difference between the two is a bit confusing.

**Comment by hanzei** [comment]
## v1 vs. v2 of github.com/nicksnyder/go-i18n

A few moth ago I added localization for [Matterpoll](https://github.com/matterpoll/matterpoll). While doing this I investigated into using v1 vs. v2 of github.com/nicksnyder/go-i18n and I want to share my thoughts here.

#### The main advantages of v2 are:
1. All default strings are kept inside the codebase. This improves the dev experience greatly. No more tapping to `en.json`.
2. Automatically extract messages from Go source code
3. Auto generation of a "have to translate" file. (See below)
4. Improved support for plural forms
5. Removed the global state
6. We are using the newest up to date version of a library, which is allways a good thing.

Please note: there are now two translation file for every language:
- `active.de.json`: This contain all translated messages
- `translate.de.json`: This contains to messages, that have to be translated. This file get auto populated every time a messages in the source changes. (This is awesome). New messages also get auto populated into this file, so you easy to track what needs to be translated. I've been using this new workflow for a few month now and it worked out amazingly. It greatly helps translator to keep track of what needs to be translated. It (to me) removed the need for a translation server.


#### Downsides of using v2 are:
1. Different version then the main server code: While I see this being a problem for core committers, it not for community members.
2. There is tooling for extracting messages from Mattermost. Why use the ones from the library?


#### Summary
I strongly think we should use v2 for the helper. Especially the greatly improved workflow for translator makes the deal for me. v1 isn't really usable without a translation server.

#### Additional information
- [Matterpoll PR that added localization](https://github.com/matterpoll/matterpoll/pull/133). I recommand checking this out, because it deals with a lot of questions and problems that come up when you are adding localization to a plugin.
- https://github.com/nicksnyder/go-i18n/blob/master/CHANGELOG.md#v2

File: plugin/helpers_i18n.go (hunk 1 above)
Comment (hmhealey) [reply plugin/helpers_i18n.go]: I guess the difference between the two settings is meaningless now that we aren't translating all of the log messages any more.

**Comment by hmhealey** [comment]
Since this is part of the server codebase, is it even possible to include both v1 and v2 of go-i18n on the same project? It seems like we'd need to migrate all of the server-side translation code to use the new version to be able to use it here. It doesn't look like we expose anything from go-i18n to the plugin, so we could easily change this in the future without affecting any plugin authors

**Comment by hanzei** [comment]
> Since this is part of the server codebase, is it even possible to include both v1 and v2 of go-i18n on the same project? It seems like we'd need to migrate all of the server-side translation code to use the new version to be able to use it here.

No, this is not needed. With go modules you can use two versions of a dependency in the same module. See https://github.com/mattermost/mattermost-server/commit/1f20ccc20e80e22587ddfbf373708c11c80d3574 for a POW.

**Comment by hanzei** [comment]
> Since this is part of the server codebase, is it even possible to include both v1 and v2 of go-i18n on the same project? It seems like we'd need to migrate all of the server-side translation code to use the new version to be able to use it here.

No, this is not needed. With go modules you can use two versions of a dependency in the same module. See https://github.com/mattermost/mattermost-server/commit/1f20ccc20e80e22587ddfbf373708c11c80d3574 for a POW.


>  It doesn't look like we expose anything from go-i18n to the plugin, so we could easily change this in the future without affecting any plugin authors

That's not possible. V2 changed the way translations are selected. You don't longer  use translationID's but instead define a [`Message`](https://godoc.org/github.com/nicksnyder/go-i18n/i18n#Message) in your code and pass this to your translation function.

**Comment by hanzei** [comment]
Hey @crspeller,

It seams like I kind of killed the discussion about this PR. This was not my intend. What are your thoughts on the v1 vs. v2 problem? If there is anything I can help you with to complete this PR, let me know.

**Comment by jasonblais** [comment]
@crspeller @hanzei Looks like this didn't get merged earlier this year (probably because focus of Extensions Team shifted to Jira plugin at the time). Wondering if we should keep this PR open or if there are any other next steps?

**Comment by crspeller** [comment]
@jasonblais This has to be redone because of the refactor of the server to use the new version of react-intl. Closing.

**Pull request closed without merging**